### PR TITLE
Feature/#17 auth layer

### LINKED ISSUE #17: store session token from login to localStorage

### PR DESCRIPTION
**Pull request by themellowpea**
# Authentication Layer Implementation

#### TLDR:

This PR introduces a robust authentication and authorization layer, featuring JWT-based authentication, role-based 
access control, and various security enhancements like stateless sessions and secure password storage in the form
of password hashing.
 

## Key Features

- **JWT Authentication**: Secure, stateless authentication using JSON Web Tokens (JWT).
- **Role-Based Access Control (RBAC)**: Managed access to API endpoints based on user roles.
- **Security Configurations**: Configured Spring Security to enforce authentication and authorization rules.
- **Comprehensive Testing**: Example tests included to validate authentication processes.

### 1. **DTO Layer**
- **LoginRequestDTO**: Validates username and password fields to ensure they are not empty.
- **UserRegistrationDTO**: Handles user registration with validation for fields like username, password, email, and roles.

### 2. **Authentication Layer**

- **AuthService**: Manages user authentication and JWT generation.
  - **Example**: 
    ```java
    String jwtToken = authService.loginUser(loginRequestDTO);
    ```

- **JwtUtil**: Utility for creating, validating, and parsing JWT tokens.
  - **Example**: 
    ```java
    String token = jwtUtil.generateToken(username, roles);
    ```

### 3. **Controller Layer**

- **AuthController**: Manages login requests and returns JWTs.
  - **Example**: 
    ```java
    @PostMapping("/login")
    public ResponseEntity<Map<String, String>> login(@RequestBody @Valid LoginRequestDTO loginDTO) {
        String jwtToken = authService.loginUser(loginDTO);
        Map<String, String> response = new HashMap<>();
        response.put("token", jwtToken);
        return ResponseEntity.ok(response);
    }
    ```

### 4. **Security Configuration**

- **SecurityConfig**: Sets up security filters, role-based access control, and password encoding.
  - **Example**: 
    ```java
    http.addFilterBefore(jwtAuthenticationFilter, UsernamePasswordAuthenticationFilter.class);
    ```

- **Role-Based Access Control**: Specifies which roles can access specific endpoints.
  - **Example**: 
    ```java
    http.authorizeHttpRequests()
        .requestMatchers("/admin/**").hasRole("ADMIN")
        .requestMatchers("/user/**").hasAnyRole("USER", "ADMIN")
        .anyRequest().authenticated();
    ```

- **PasswordEncoder**: Configures `BCryptPasswordEncoder` for secure password hashing.
  - **Example**: 
    ```java
    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }
    ```

### 5. **Testing**

- **AuthControllerTest**: Includes tests for successful login, failure scenarios, and validation errors.
  - **Example**: 
    ```java
    mockMvc.perform(post("/auth/login").contentType(MediaType.APPLICATION_JSON)
            .content("{\"username\":\"testuser1\", \"password\":\"valid_password\"}"))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.token").isNotEmpty());
    ```

### CODE REVIEW

**Review by trevorlana — APPROVED** [review]
Looks good

**Review by revature-ben-m — APPROVED** [review]
Very nice!
